src/ports.js

```diff
@@ -0,0 +1,440 @@
+
+                this._preparePorts();
+            }, this);
+        },
+
+        hasPorts: function() {
+
+            return this.getPorts().length > 0;
+        },
+
+        getPorts: function() {
+
+            return _.cloneDeep(this.prop('ports/items')) || [];
+        },
+
+        getPort: function(id) {
+
+            return _.find(this.getPorts(), function(port) {
+                return port.id && port.id === id;
+            });
+        },
+
+        getPortIndex: function(id) {
+
+            return _.findIndex(this.getPorts(), function(port) {
+                return port.id && port.id === id;
+            });
+        },
+
+        addPort: function(port, opt) {
+
+            var ports = _.clone(this.prop('ports/items')) || [];
+            ports.push(port || {});
+            this.prop('ports/items', ports, opt);
+
+            return this;
+        },
+
+        _validatePorts: function() {
+
+            var portsAttr = this.get('ports') || {};
+
+            var errorMessages = [];
+            portsAttr = portsAttr || {};
+            var ports = portsAttr.items || [];
+
+            _.each(ports, function(p) {
+                if (!p.id) {
+                    p.id = joint.util.uuid();
+                }
+            }, this);
+
+            if (_.uniq(ports, 'id').length !== ports.length) {
+                errorMessages.push('Element: found id duplicities in ports.');
+            }
+
+            return errorMessages;
+        },
+
+        addPorts: function(ports, opt) {
+
+            if (ports.length) {
+                this.prop('ports/items', this.getPorts().concat(ports), opt);
+            }
+
+            return this;
+        },
+
+        removePort: function(port, opt) {
+
+            opt = opt || {};
+            var ports = this.getPorts();
+
+            var index = _.isString(port) ? _.findIndex(ports, this.getPort(port)) : _.findIndex(ports, port);
+
+            if (index !== -1) {
+                ports.splice(index, 1);
+                opt.rewrite = true;
+                this.prop('ports/items', ports, opt);
+            }
+
+            return this;
+        },
+
+        /**
+         * @private
+         */
+        _preparePorts: function() {
+
+            var err = this._validatePorts();
+
+            if (err.length > 0) {
+                this.set('ports', this.previous('ports'));
+                throw new Error(err.join(' '));
+            }
+
+            this.portData = new PortData(this.get('ports'));
+        }
+    });
+
+    _.extend(joint.dia.ElementView.prototype, {
+
+        portContainerMarkup: '<g class="port"/>',
+        portMarkup: '<circle class="base-port" r="10" fill="#000000"/>',
+        portLabelMarkup: '<text class="label"/>',
+
+        /**
+         * @private
+         */
+        _initializePorts: function() {
+
+            /** @type {Object<string, {portElement: V, portLabelElement: V}>} */
+            this._portElementsCache = {};
```

Comment:
Could this be set via the extend method a few lines above - like this?
```js
    _.extend(joint.dia.ElementView.prototype, {

        portContainerMarkup: '<g class="port"/>',
        portMarkup: '<circle class="base-port" r="10" fill="#000000"/>',
        portLabelMarkup: '<text class="label"/>',

        _portElementsCache: {},
        // ..
```

I like to know all the keys that are used on the object's namespace.